federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java

```diff
@@ -237,16 +233,33 @@ public boolean supportsCredentialAuthenticationFor(String type) {
 
              for (LDAPObject ldapUser : ldapObjects) {
                  String ldapUsername = LDAPUtils.getUsername(ldapUser, this.ldapIdentityStore.getConfig());
-                 if (session.userLocalStorage().getUserByUsername(ldapUsername, realm) == null) {
+                 UserModel localUser = session.userLocalStorage().getUserByUsername(ldapUsername, realm);
+                 if (localUser == null) {
                      UserModel imported = importUserFromLDAP(session, realm, ldapUser);
                      searchResults.add(imported);
+                 } else if (shouldUserAttributeBeAlwaysReadFromLdap(realm, attrName)) {
+                     searchResults.add(proxy(realm, localUser, ldapUser));
```

Comment:
I think it will be good to remove method "shouldUserAttributeBeAlwaysReadFromLdap" (and any new things related to that added in this PR) and just always add the proxy to the results.

Regarding this, it may be also good to update a bit method UserStorageManager.query and make sure that duplicates are not returned. As currently UserStorageMAnager.query returns stuff from multiple providers, so it can probably return duplicated stuff in some cases (EG. in case that LDAP user was imported and attribute "foo" with value "bar" is available in both the DB and LDAP)

Reply:
In which case would we consider two users from different stores to be duplicates?
They may be duplicates when user from LDAP has been imported to local store. Then LDAP user and user from local store could be considered duplicates.
Should we treat users from (let say) a custom storage provider and the LDAP provider to be duplicates (e.g. when usernames are equal)? 

And if we have duplicates, which user model should be removed from the result set?
Is it always the one from the local store, because the one from LDAP provider will always proxy the local one anyways?